chore(docs): Update cache docs (#17420)
grafana/loki: docs/sources/operations/caching.md

---
# Configure caches to speed up queries

Loki supports two types of caching for query results and chunks to speed up query performance and reduce calls to storage layer.
This sections describes the recommended Memcached configuration to enable caching for chunks and query results.

#### Results cache

> The results cache stores the results for index-stat, instant-metric, label and volume queries and it supports negative caching for log queries. It is sometimes called frontend cache in some configurations. For details of each supported request type please see [Components section](https://grafana.com/docs/loki/<LOKI_VERSION>//get-started/components)

Reviewer: ```suggestion
The results cache stores the results for index-stat, instant-metric, label and volume queries and it supports negative caching for log queries. It is sometimes called frontend cache in some configurations. For details of each supported request type please see [Components section](https://grafana.com/docs/loki/<LOKI_VERSION>//get-started/components).
```

> It's consulted by query-frontends to be used in subsequent queries. If the cached results are incomplete, the query frontend calculates the required sub-queries and sends them further to be executed in queriers, then also cache those results.

Reviewer: ```suggestion
It's consulted by query-frontends to be used in subsequent queries. If the cached results are incomplete, the query frontend calculates the required sub-queries and sends them further along to be executed in queriers, then also caches those results.
```

> To orchestrate all above, the results cache uses a query hash as the key that is computed and stored in the headers.

Reviewer: ```suggestion
To orchestrate all of the above, the results cache uses a query hash as the key that is computed and stored in the headers.
```

The index lookup cache only supported the legacy BoltDB index storage and is configured to be in-memory by default.
Since moving to the TSDB indexes the attached disks/persistant volumes are utilised as cache and in-memory index lookup cache is obsolete.

#### Chunks cache

> The cache chunks using the `chunkRef` as the cache key, which is the unique reference to a chunk when it's cut in ingesters.

Reviewer: ```suggestion
The chunks are cached using the `chunkRef` as the cache key, which is the unique reference to a chunk when it's cut in the Loki ingesters.
```

It's consulted by queriers each time a set of `chunkRef`s are calculated to serve the query, before going to storage layer.

> Query results are significantly smaller compared to chunks. As the Loki cluster get bigger in ingested volume, results cache can continue to perform whereas chunks cache will need to grow in proportion to demand more memory.

Reviewer: ```suggestion
Query results are significantly smaller compared to chunks. As the Loki cluster gets bigger in ingested volume, the results cache can continue to perform, whereas the chunks cache will need to grow in proportion to demand more memory.
```

#### Results cache
The results cache stores the results for index-stat, instant-metric, label and volume queries and it supports negative caching for log queries. It is sometimes called frontend cache in some configurations. For details of each supported request type, refer to the [Components section](https://grafana.com/docs/loki/<LOKI_VERSION>//get-started/components).
The results cache is consulted by query-frontends to be used in subsequent queries. If the cached results are incomplete, the query frontend calculates the required sub-queries and sends them further along to be executed in queriers, then also caches those results.
To orchestrate all of the above, the results cache uses a query hash as the key that is computed and stored in the headers.

The index lookup cache only supports the legacy BoltDB index storage and is configured to be in-memory by default.
Since moving to the TSDB indexes the attached disks/persistent volumes are utilised as cache and in-memory index lookup cache is obsolete.

#### Chunks cache
The chunks are cached using the `chunkRef` as the cache key, which is the unique reference to a chunk when it's cut in the Loki ingesters.

> The chunk cache is consulted by queriers each time a set of `chunkRef`s are calculated to serve the query, before going to storage layer.

Reviewer 2: ```suggestion
The chunk cache is consulted by queriers each time a set of `chunkRef`s are calculated to serve the query, before going to the storage layer.
```

Query results are significantly smaller compared to chunks. As the Loki cluster gets bigger in ingested volume, the results cache can continue to perform, whereas the chunks cache will need to grow in proportion to demand more memory.

> To be able to support the growing needs of a cluster, in 2023 we introduced support for memcached-extstore. Exstore is an additional feature on Memcached which supports attaching SSD disks to memcached pods to maximize their capacity.

Reviewer 2: ```suggestion
To be able to support the growing needs of a cluster, in 2023 we introduced support for memcached-extstore. Extstore is an additional feature on Memcached which supports attaching SSD disks to memcached pods to maximize their capacity.
```

> Please see this [blog post](https://grafana.com/blog/2023/08/23/how-we-scaled-grafana-cloud-logs-memcached-cluster-to-50tb-and-improved-reliability/) on Loki's experience with memcached-extstore for our SaaS offfering.

Reviewer 2: ```suggestion
Please see this [blog post](https://grafana.com/blog/2023/08/23/how-we-scaled-grafana-cloud-logs-memcached-cluster-to-50tb-and-improved-reliability/) on Loki's experience with memcached-extstore for our SaaS offfering, Grafana Cloud.
```

For more information on how to tune memcached-extstore please consult the open source [memcached documentation](https://docs.memcached.org/advisories/grafanaloki/).

## Before you begin

- It is recommended to deploy separate Memcached type as separate components (`memcached_frontend` and `memcached_chunks`).
- As of 2025-02-06, the `memcached:1.6.32-alpine` version of the library is recommended.
- Consult the Loki ksonnet [memcached](https://github.com/grafana/loki/blob/main/production/ksonnet/loki/memcached.libsonnet) deployment and the ksonnet [memcached library](https://github.com/grafana/jsonnet-libs/tree/master/memcached).
- Index caching is not required for the [TSDB](https://grafana.com/docs/loki/<LOKI_VERSION>/operations/storage/tsdb/#index-caching-not-required) index format.

> - Please see [Size the cluster](https://grafana.com/docs/loki/<LOKI_VERSION>/setup/size/)  page for recommendations on scaling the cache.

Reviewer 2: ```suggestion
- For recommendations on scaling the cache, refer to the [Size the cluster](https://grafana.com/docs/loki/<LOKI_VERSION>/setup/size/) page.
```
